File: backend/tissue-core/src/main/java/com/tissue/shared/enums/ColorType.java
package com.tissue.shared.enums;

import java.util.Arrays;
import java.util.List;
import java.util.Random;
import java.util.Set;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

/**
 * Colors based from Textual's
 * <a href="https://textual.textualize.io/api/color/#textual.color.Color">named colors</a>.
 *
 * <p>All ANSI colors are provided, other colors are partially provided.
 * Using the HexCode and display name will not be needed in most cases. Let the client (Textual) render
 * the color based on the enum value.
 */
@Getter
@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
public enum ColorType {
    // ANSI 16 colors
    ANSI_BLACK("#000000", "ANSI Black"),
    ANSI_RED("#800000", "ANSI Red"),
    ANSI_GREEN("#008000", "ANSI Green"),
    ANSI_YELLOW("#808000", "ANSI Yellow"),
    ANSI_BLUE("#000080", "ANSI Blue"),
    ANSI_MAGENTA("#800080", "ANSI Magenta"),
    ANSI_CYAN("#008080", "ANSI Cyan"),
    ANSI_WHITE("#C0C0C0", "ANSI White"),
    ANSI_BRIGHT_BLACK("#808080", "ANSI Gray"),
    ANSI_BRIGHT_RED("#FF0000", "ANSI Bright Red"),
    ANSI_BRIGHT_GREEN("#00FF00", "ANSI Bright Green"),
    ANSI_BRIGHT_YELLOW("#FFFF00", "ANSI Bright Yellow"),
    ANSI_BRIGHT_BLUE("#0000FF", "ANSI Bright Blue"),
    ANSI_BRIGHT_MAGENTA("#FF00FF", "ANSI Bright Magenta"),
    ANSI_BRIGHT_CYAN("#00FFFF", "ANSI Bright Cyan"),
    ANSI_BRIGHT_WHITE("#FFFFFF", "ANSI Bright White"),

    // Extended colors
    PINK("#FFC0CB", "Pink"),
    MAROON("#800000", "Maroon"),
    RED("#FF0000", "Red"),
    ORANGERED("#FF4500", "Orange Red"),
    DARKORANGE("FF8C00", "Dark Orange"),
    LIMEGREEN("#32CD32", "Lime Green"),
    LIGHTGREEN("90EE90", "Light Green"),
    LIGHTYELLOW("#FFFFE0", "Light Yellow"),
    MEDIUMBLUE("#0000CD", "Medium Blue"),
    MIDNIGHTBLUE("#191970", "Midnight Blue"),
    INDIGO("#4B0082", "Indigo"),
    MAGENTA("#FF00FF", "Magenta"),
    BROWN("#A52A2A", "Brown"),
    TAN("#D2B48C", "Tan");

    private static final Random RANDOM = new Random();
    private final String hexCode;
    private final String displayName;

    public static ColorType getRandomUnusedColor(Set<ColorType> usedColors) {
        List<ColorType> availableColors = Arrays.stream(ColorType.values())
                .filter(color -> !usedColors.contains(color))
                .toList();

        if (availableColors.isEmpty()) {
            availableColors = Arrays.asList(ColorType.values());
        }

        return availableColors.get(RANDOM.nextInt(availableColors.size()));
    }

    public static ColorType getRandomColor() {
        return ColorType.values()[RANDOM.nextInt(ColorType.values().length)];
    }
}

File: client/tui/src/tissue/widgets/color_type.py
"""Turn a ColorType enum name (e.g. "PINK", "ANSI_RED") into a `#rrggbb` hex.

Uses Textual's color parser because it understands the ANSI names that Rich
does not.
"""

from __future__ import annotations

from textual.color import Color as TextualColor
from textual.color import ColorParseError as TextualColorParseError

# ColorType enum names, in enum order, matching the generated color validators
# that mirror the backend ColorType.java enum (the one place colors are defined).
COLOR_NAMES: list[str] = [
    "ANSI_BLACK",
    "ANSI_RED",
    "ANSI_GREEN",
    "ANSI_YELLOW",
    "ANSI_BLUE",
    "ANSI_MAGENTA",
    "ANSI_CYAN",
    "ANSI_WHITE",
    "ANSI_BRIGHT_BLACK",
    "ANSI_BRIGHT_RED",
    "ANSI_BRIGHT_GREEN",
    "ANSI_BRIGHT_YELLOW",
    "ANSI_BRIGHT_BLUE",
    "ANSI_BRIGHT_MAGENTA",
    "ANSI_BRIGHT_CYAN",
    "ANSI_BRIGHT_WHITE",
    "PINK",
    "MAROON",
    "RED",
    "ORANGERED",
    "DARKORANGE",
    "LIMEGREEN",
    "LIGHTGREEN",
    "LIGHTYELLOW",
    "MEDIUMBLUE",
    "MIDNIGHTBLUE",
    "INDIGO",
    "MAGENTA",
    "BROWN",
    "TAN",
]


def _rgb(color: str | None) -> tuple[int, int, int] | None:
    if not color:
        return None
    # The usual value is the ColorType enum NAME ("ansi_bright_blue", "limegreen"),
    # but also accept a raw hex ("#ff4500", from a theme variable) and a display
    # name ("ANSI Bright Blue"). Try the value three ways and use whichever
    # Textual's parser recognizes.
    #   - as-is
    #   - spaces turned into underscores
    #   - spaces removed
    base = color.strip().lower()
    for candidate in (base, base.replace(" ", "_"), base.replace(" ", "")):
        try:
            return TextualColor.parse(candidate).rgb
        except TextualColorParseError:
            continue
    return None


def color_hex(color: str | None) -> str:
    """A ``#rrggbb`` hex for a ColorType enum name.

    Returns "" when it is empty or not a color Textual knows.
    """
    rgb = _rgb(color)
    return f"#{rgb[0]:02x}{rgb[1]:02x}{rgb[2]:02x}" if rgb else ""


def color_fg(color: str | None) -> str:
    """A readable foreground (#000000 / #ffffff) for text drawn ON `color`.

    Black on light backgrounds, white on dark, or "" when there is no color.
    """
    rgb = _rgb(color)
    if rgb is None:
        return ""
    r, g, b = rgb
    # perceived luminance (ITU-R BT.601), above ~55% means use black text
    return "#000000" if (0.299 * r + 0.587 * g + 0.114 * b) > 140 else "#ffffff"


def chip_style(color: str | None) -> str:
    """A Rich style ``"<fg> on <bg>"`` filling `color` as a solid pill with a
    readable foreground, or "" when there is no color."""
    bg = color_hex(color)
    if not bg:
        return ""
    return f"{color_fg(color)} on {bg}"
